lib/Transforms/IPO/StructFieldAccessInfo.cpp

```diff
@@ -0,0 +1,349 @@
+// StructFieldCacheAnalysis pass.
+//
+//===------------------------------------------------------------------------===//
+
+#include "StructFieldCacheAnalysisImpl.h"
+#include "llvm/IR/CallSite.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Operator.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "struct-analysis"
+#define DEBUG_TYPE_IR "struct-analysis-IR"
+#define DEBUG_TYPE_STATS "struct-analysis-detailed-stats"
+
+static cl::opt<unsigned> HistogramSizeForStats(
+    "struct-analysis-number-buckets", cl::init(10), cl::Hidden,
+    cl::desc("Number of buckets used to analyze struct hotness"));
+
+static cl::opt<unsigned> HotnessCutoffRatio(
+    "struct-analysis-hotness-cutoff", cl::init(0), cl::Hidden,
+    cl::desc("Filter out structs that is colder than a ratio of maximum "
+             "hotness, should be a percentage."));
+
+// Functions for StructFieldAccessInfo
+// Add a record of load/store Instruction I accessing field FieldNum in the map
+void StructFieldAccessInfo::addFieldAccessNum(const Instruction *I,
+                                              FieldNumType FieldNum) {
+  assert(isa<LoadInst>(I) || isa<StoreInst>(I)); // Only loads and stores
+  assert(LoadStoreFieldAccessMap.find(I) == LoadStoreFieldAccessMap.end());
+  LoadStoreFieldAccessMap[I] = FieldNum;
+}
+
+// Add a record of call/invoke Instruction I using field FieldNum as its
+// argument Arg F is the defition of the callee function
+void StructFieldAccessInfo::addFieldAccessNum(const Instruction *I,
+                                              const Function *F, unsigned Arg,
+                                              FieldNumType FieldNum) {
+  assert(isa<CallInst>(I) || isa<InvokeInst>(I)); // Only calls and invokes
+  if (F->isDeclaration())
+    // Give up if the function only has declaration
+    return;
+  // Create a new record FunctionCallInfo if this Arg->FieldNum pair of the
+  // Instruction is the first pair. Otherwise, insert a new pair of
+  // Arg->FieldNum to an existing FunctionCallInfo object
+  if (CallInstFieldAccessMap.find(I) == CallInstFieldAccessMap.end()) {
+    CallInstFieldAccessMap[I] = new FunctionCallInfo(F, Arg, FieldNum);
+  } else {
+    auto *CallSite = CallInstFieldAccessMap[I];
+    assert(CallSite->FunctionDeclaration == F);
+    CallSite->insertCallInfo(Arg, FieldNum);
+  }
+}
+
+Optional<FieldNumType>
+StructFieldAccessInfo::getAccessFieldNum(const Instruction *I) const {
+  Optional<FieldNumType> ret;
+  auto it = LoadStoreFieldAccessMap.find(I);
+  if (it != LoadStoreFieldAccessMap.end()) {
+    ret = it->second;
+  }
+  return ret;
+}
+
+FieldNumType
+StructFieldAccessInfo::calculateFieldNumFromGEP(const User *U) const {
+  DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Calculating field number from GEP: "
+                                        << *U << "\n");
+  // Operand 0 should be a pointer to the struct
+  assert(isa<GetElementPtrInst>(U) || isa<GEPOperator>(U) ||
```

Comment:
Yeah... I think so, I was over-designed on the assertions to make sure everything's right. It should be OK to remove it. I'll update it in the next pull request.